File: tests/journal.pkg/courier_full.py
#!/usr/bin/env python3
# -*- Python -*-
# -*- coding: utf-8 -*-
#
# michael a.g. aïvázis <michael.aivazis@para-sim.com>
# (c) 1998-2026 all rights reserved


def test():
    """
    A far end that does not keep up costs records, never time; the loss is counted and reported
    """
    # externals
    import os

    # access
    import journal

    # make a pipe; nobody reads it for a while
    reader, writer = os.pipe()
    # make a courier on its write end and install it
    courier = journal.courier(descriptor=writer)
    journal.chronicler.device = courier
    # the channel
    channel = journal.info("test.courier.full")

    # a line long enough to fill any pipe buffer in a few hundred entries
    filler = "x" * 1024
    # the number of entries to attempt
    attempts = 1024
    # log them all without reading; each call must return, whatever the state of the pipe
    for _ in range(attempts):
        # log
        channel.log(filler)
    # every attempt was stamped, and so was every report of drops that found room to go out
    # between refusals, so the sequence runs at least as far as the attempts
    assert courier.seq >= attempts
    # some were lost
    assert courier.dropped > 0

    # drain the pipe; the read end must not block either
    os.set_blocking(reader, False)
    # collect what made it through
    received = b""
    # read until there is nothing left
    while True:
        # attempt to
        try:
            # read a chunk
            chunk = os.read(reader, 64 * 1024)
        # if the pipe is empty
        except BlockingIOError:
            # done
            break
        # if the pipe is empty
        if not chunk:
            # done
            break
        # accumulate
        received += chunk

    # log once more; the far end has caught up, so this one goes out, preceded by the report
    channel.log("after")
    # the drops were reported
    assert courier.dropped == 0
    # read the rest
    while True:
        # attempt to
        try:
            # read a chunk
            chunk = os.read(reader, 64 * 1024)
        # if the pipe is empty
        except BlockingIOError:
            # done
            break
        # if the pipe is empty
        if not chunk:
            # done
            break
        # accumulate
        received += chunk

    # split into records; every line is whole, since records are never torn
    lines = received.splitlines(keepends=True)
    assert all(line.endswith(b"\n") for line in lines)
    # decode
    records = [journal.record.decode(line) for line in lines]
    # the last two are the report and the entry that prompted it
    notice, after = records[-2:]
    # the report is on the courier's own channel
    assert notice.channel == "journal.courier"
    assert notice.severity == "warning"
    # and the entry that prompted it follows
    assert after.page == ["after"]
    # the sequence numbers of the records that arrived are increasing
    sequence = [int(record.notes["seq"]) for record in records]
    assert sequence == sorted(sequence)
    assert len(set(sequence)) == len(sequence)
    # the reports of drops, each accounting for at least one loss
    notices = [record for record in records if record.channel == "journal.courier"]
    assert all(int(record.notes["dropped"]) > 0 for record in notices)
    reported = sum(int(record.notes["dropped"]) for record in notices)
    # the entries that made it out are the records that are not reports
    assert courier.shipped == len(records) - len(notices)
    # every attempt, the final entry included, was either shipped or reported lost
    assert courier.shipped + reported == attempts + 1
    # and the gaps in the sequence account for exactly the losses
    assert sequence[-1] - len(records) == reported

    # clean up
    courier.close()
    os.close(reader)

    # all done
    return


# main
if __name__ == "__main__":
    # tell journal to stay away from the bindings
    journal_no_libjournal = True
    # run the test
    test()


# end of file

File: tests/journal.lib/courier_full.cc
// -*- C++ -*-
// -*- coding: utf-8 -*-
//
// michael a.g. aïvázis <michael.aivazis@para-sim.com>
// (c) 1998-2026 all rights reserved


// get the journal
#include <pyre/journal.h>
// support
#include <cassert>
#include <cerrno>
#include <fcntl.h>
#include <unistd.h>


// alias the types
using courier_t = pyre::journal::courier_t;
using chronicler_t = pyre::journal::chronicler_t;
using info_t = pyre::journal::info_t;


// read everything a non-blocking descriptor has to offer right now
static auto
slurp(int descriptor) -> std::string
{
    // the pile
    std::string pile;
    // a chunk
    char buffer[64 * 1024];
    // for as long as there is something
    while (true) {
        // read a chunk
        auto got = ::read(descriptor, buffer, sizeof(buffer));
        // if the pipe is empty, or drained
        if (got <= 0) {
            // done
            break;
        }
        // accumulate
        pile.append(buffer, got);
    }
    // all done
    return pile;
}


// split a byte string into lines, reporting whether the last one was terminated
static auto
lines(const std::string & data) -> std::vector<std::string>
{
    // the pile
    std::vector<std::string> pile;
    // the start of the current line
    std::string::size_type start = 0;
    // for as long as there are newlines
    while (true) {
        // find the next one
        auto end = data.find('\n', start);
        // if there isn't one
        if (end == std::string::npos) {
            // whatever is left is a partial line, which must not happen
            assert(start == data.size());
            // done
            break;
        }
        // save the line
        pile.push_back(data.substr(start, end - start));
        // and move past it
        start = end + 1;
    }
    // all done
    return pile;
}


// pull a numeric field out of a record
static auto
field(const std::string & record, const std::string & key) -> long
{
    // find the key
    auto at = record.find(key);
    assert(at != std::string::npos);
    // and read the number that follows it
    return std::stol(record.substr(at + key.size()));
}


// a far end that does not keep up costs records, never time; the loss is counted and reported
int
main()
{
    // make a pipe; nobody reads it for a while
    int pipes[2];
    assert(::pipe(pipes) == 0);
    // make a courier on its write end and install it
    auto courier = std::make_shared<courier_t>(pipes[1]);
    chronicler_t::device(courier);

    // a line long enough to fill any pipe buffer in a few hundred entries
    std::string filler(1024, 'x');
    // the number of entries to attempt
    const size_t attempts = 1024;
    // log them all without reading; each call must return, whatever the state of the pipe
    for (size_t i = 0; i < attempts; ++i) {
        info_t channel("test.courier.full");
        channel << filler << pyre::journal::endl;
    }
    // every attempt was stamped
    assert(courier->seq() == attempts);
    // some were lost
    assert(courier->dropped() > 0);
    // and the books balance
    assert(courier->shipped() + courier->dropped() == attempts);
    // the lost ones count what was owed
    auto owed = courier->dropped();

    // drain the pipe; the read end must not block either
    ::fcntl(pipes[0], F_SETFL, ::fcntl(pipes[0], F_GETFL, 0) | O_NONBLOCK);
    auto received = slurp(pipes[0]);
    // log once more; the far end has caught up, so this one goes out, preceded by the report
    info_t after("test.courier.full");
    after << "after" << pyre::journal::endl;
    // the drops were reported
    assert(courier->dropped() == 0);
    // read the rest
    received += slurp(pipes[0]);

    // split into records; every line is whole, since records are never torn
    auto records = lines(received);
    assert(records.size() >= 2);
    // the last two are the report and the entry that prompted it
    const auto & notice = records[records.size() - 2];
    const auto & last = records[records.size() - 1];
    // the report is on the courier's own channel
    assert(notice.find("\"channel\":\"journal.courier\"") != std::string::npos);
    assert(notice.find("\"severity\":\"warning\"") != std::string::npos);
    assert(notice.find("\"sink\":\"alert\"") != std::string::npos);
    // it carries the count in its notes
    assert(notice.find("\"dropped\":\"" + std::to_string(owed) + "\"") != std::string::npos);
    // and the entry that prompted it follows
    assert(last.find("\"page\":[\"after\"]") != std::string::npos);
    // the sequence numbers of the records that arrived are increasing
    long previous = 0;
    for (const auto & record : records) {
        auto seq = field(record, "\"seq\":");
        assert(seq > previous);
        previous = seq;
    }
    // and the gaps account for exactly the records owed
    assert(field(last, "\"seq\":") - static_cast<long>(records.size()) == static_cast<long>(owed));

    // clean up
    courier->close();
    ::close(pipes[0]);

    // all done
    return 0;
}


// end of file
